Ark.Tools.FtpClient.SftpClient/SftpClientConnection.cs

```diff
@@ -87,19 +108,117 @@ public override async Task UploadFileAsync(string path, byte[] content, Cancella
+                Timeout = TimeSpan.FromMinutes(5),
+                RetryAttempts = 2
+            };
+
+            return connrectionInfo;
+        }
+
+        private Renci.SshNet.SftpClient _getSFtpClientWithCertificate()
         {
-            var connInfo = new ConnectionInfo(Uri == null ? Host : Uri.Host, Uri == null ? Port : Uri.Port, Credentials.UserName, new PasswordAuthenticationMethod(Credentials.UserName, Credentials.Password));
-            connInfo.Timeout = TimeSpan.FromMinutes(5);
-            connInfo.RetryAttempts = 2;
-           
-            return new Renci.SshNet.SftpClient(connInfo)
+
+#if NETSTANDARD2_0
+            throw new NotSupportedException($"ClientCertificate does not support X509 Certificate in NETCORE2.0");
+#endif
+            var connInfo = _getConnectionInfo();
+
+            using var cert = _ftpConfig.ClientCertificate;
+
+            string keyExchangeAlgorithm = null;
+            byte[] privateKeyBytes = null;
+            string privateKeyPemString;
+            bool isKeyNull = false;
+
+            switch (cert.PublicKey.Oid.Value)
+            {
+                case _rsa:
+                    {
+                        using RSA rsaKey = cert.GetRSAPrivateKey();
+
+                        keyExchangeAlgorithm = rsaKey.KeyExchangeAlgorithm;
+                        if (rsaKey != null)
+                            privateKeyBytes = rsaKey.ExportRSAPrivateKey();
+                        else
+                            isKeyNull = true;
+                        break;
+                    }
+                case _dsa:
+                    {
+                        using DSA dsaKey = cert.GetDSAPrivateKey();
+
+                        keyExchangeAlgorithm = dsaKey.KeyExchangeAlgorithm;
+                        if (dsaKey != null)
+                            privateKeyBytes = dsaKey.ExportPkcs8PrivateKey();
+                        else
+                            isKeyNull = true;
+                        break;
+                    }
+                case _ecdsa:
+                    {
+                        using ECDsa ecdsaKey = cert.GetECDsaPrivateKey();
+
+                        keyExchangeAlgorithm = ecdsaKey.KeyExchangeAlgorithm;
+                        if (ecdsaKey != null)
+                            privateKeyBytes = ecdsaKey.ExportPkcs8PrivateKey();
+                        else
+                            isKeyNull = true;
+                        break;
+                    }
+                default:
+                    throw new NotSupportedException($"ClientCertificate does not support the given algorithm {cert.PublicKey.Oid.FriendlyName}");
+            }
+
+            if (isKeyNull)
+                throw new ArgumentNullException($"ClientCertificate has a null Key");
+
+#if NET5_0_OR_GREATER
+            var privateKeyPem = PemEncoding.Write($"{keyExchangeAlgorithm} PRIVATE KEY", privateKeyBytes);
+            privateKeyPemString = new string(privateKeyPem);
```

Comment:
strange string ctor. What PemEncoding.Write returns?

Reply:
char[], this is why I used the ctor for string